src/palace/manager/celery/tasks/reaper.py

```diff
@@ -0,0 +1,239 @@
+    for 60 days), and must not be associated with one of the
+    patron's active loans or holds.
+    """
+    cutoff = utc_now() - timedelta(days=60)
+
+    restrictions = []
+    for t in Loan, Hold:
+        active_subquery = (
+            select(Annotation.id)
+            .join(t, t.patron_id == Annotation.patron_id)
+            .join(
+                LicensePool,
+                and_(
+                    LicensePool.id == t.license_pool_id,
+                    LicensePool.identifier_id == Annotation.identifier_id,
+                ),
+            )
+        )
+        restrictions.append(~Annotation.id.in_(active_subquery))
+
+    deletion_query = delete(Annotation).where(
+        Annotation.timestamp < cutoff,
+        Annotation.motivation == Annotation.IDLING,
+        *restrictions,
+    )
+
+    with task.transaction() as session:
+        rows_removed = _execute_delete(session, deletion_query)
+    task.log.info(f"Deleted {pluralize(rows_removed, 'outdated idling annotation')}.")
+
+
+@shared_task(queue=QueueNames.default, bind=True)
+def hold_reaper(task: Task, batch_size: int = 100) -> None:
+    """
+    Remove seemingly abandoned holds from the database.
+    """
+    cutoff = utc_now() - timedelta(days=365)
+    analytics_service = task.services.analytics.analytics()
+    query = (
+        select(Hold)
+        .where(Hold.start < cutoff, or_(Hold.end == None, Hold.end < utc_now()))
+        .order_by(Hold.id)
+        .limit(batch_size)
+    )
+    events_to_be_logged = []
+    with task.transaction() as session:
+        holds = session.execute(query).scalars().all()
+        for hold in holds:
+            events_to_be_logged.append(
+                AnalyticsEventData.create(
+                    library=hold.library,
+                    license_pool=hold.license_pool,
+                    event_type=CirculationEvent.CM_HOLD_EXPIRED,
+                    patron=hold.patron,
+                )
+            )
+            session.delete(hold)
+
+    count = len(holds)
+    task.log.info(f"Deleted {pluralize(count, 'expired hold')}.")
+
+    with task.transaction() as session:
+        for event in events_to_be_logged:
+            analytics_service.collect(event=event, session=session)
+
+    if count == batch_size:
+        task.log.info("There may be more holds to delete. Re-queueing the reaper.")
+        raise task.replace(hold_reaper.s(batch_size=batch_size))
+
+
+@shared_task(queue=QueueNames.default, bind=True)
+def loan_reaper(task: Task) -> None:
+    """
+    Remove expired and abandoned loans from the database.
+    """
+
+    now = utc_now()
+    deletion_query = delete(Loan).where(
+        Loan.license_pool_id == LicensePool.id,
+        LicensePool.unlimited_access == False,  # type: ignore[comparison-overlap]
+        LicensePool.open_access == False,
+        or_(
+            Loan.end < now,
+            and_(Loan.start < now - timedelta(days=90), Loan.end == None),
+        ),
```

Comment:
This query drops the concept of "source of truth" integrations.

Previously the only source of truth integration we had was OPDS for distributors, which only supports unlimited access items, so these items are filtered out by the `LicensePool.unlimited_access == False` check, so no need to complicate the query by filtering on the collection protocol.

So this query should be effectively equivalent to what we had before.